### DISCUSSION TITLE
ContextMenu.The issue of updating data disappearing

### DISCUSSION THREAD

**Discussion by wosiyourfather** [AvaloniaUI/Avalonia#12393] [Q&A]
https://github.com/AvaloniaUI/Avalonia/assets/53590491/5af31d6a-f664-4799-b630-b3a418519ae2

**Comment by wosiyourfather** [comment]
help~~~

**Comment by wosiyourfather** [comment]
Data update, the menu will disappear

**Reply by timunie** [reply]
Are you using the MVVM pattern or do you use code-behind to update the Menu?. Please show us your code to generate the menu in order to check.

**Reply by wosiyourfather** [reply]
MVVM <ListBox Name="curList" GotFocus="Sssssss_GotFocus"     ItemsSource="{Binding nodeIO2}" >
									<ListBox.ItemsPanel>
									<ItemsPanelTemplate>
										<WrapPanel/>
									</ItemsPanelTemplate>
								</ListBox.ItemsPanel>
									<ListBox.Styles>
										<Style Selector="Label.Nodes">
										<Setter Property="Width" Value="{Binding stylse.Width,Mode=TwoWay}"/>
										<Setter Property="Height" Value="{Binding stylse.Height,Mode=TwoWay}"/>
										</Style>
									</ListBox.Styles>
								<ListBox.ItemTemplate>
									<DataTemplate>
								 <Border Grid.ColumnSpan="2"  Grid.RowSpan="3" Background="#FFF5EFEF" BorderThickness="1" CornerRadius="3" BorderBrush="#999999">
									 <Border.ContextMenu>
										 <ContextMenu  Name="cd">
											 <MenuItem Name="cdd" Header="kkxiaoguo"></MenuItem>
											 <MenuItem Name="cddw" Header="kkxiaoguo"></MenuItem>
											 <MenuItem Name="cddwe" Header="kkxiaoguo"></MenuItem>
										 </ContextMenu>
									 </Border.ContextMenu>
									<Grid Classes="Node" Width="200" Height="200" >
										
										<Grid.ColumnDefinitions >
											<ColumnDefinition Width="*"/>
											<ColumnDefinition Width="*"/>
										</Grid.ColumnDefinitions>
										<Grid.RowDefinitions>
											<RowDefinition Height="*"></RowDefinition>
											<RowDefinition Height="*"></RowDefinition>
											<RowDefinition Height="*"></RowDefinition>
										    <RowDefinition Height="*"></RowDefinition>
										</Grid.RowDefinitions>
										<Label Grid.Column="0" Classes="Nodes"    FontSize="12" Grid.Row="0"  Name="lbname" Content="{Binding name+name}"></Label>
										<Label Grid.Column="1" Classes="Nodes"   FontSize="12" Grid.Row="0"  Name="lbpower" Content="{Binding power,StringFormat={}{0}kW}"></Label>
										
										<Label Grid.Column="1"  Classes="Nodes"  FontSize="12" Grid.Row="1"  Name="lbwindspeed" Content="{Binding winspeed,StringFormat={}{0}m/s}"></Label>
										<Label Grid.Column="1" Classes="Nodes"  FontSize="12" Grid.Row="2"  Name="lbproduct" Content="{Binding product,StringFormat={}{0}Mwh}"></Label>
										
										
									</Grid>

**Reply by timunie** [reply]
Ah okay, so the ContextMenu is attached to a border. The border should have a Background set. `Transparent` is okay, but `null` (which is the default) is not okay as the mouse cannot hit the control in that case.

**Reply by wosiyourfather** [reply]
I have already set the background
![1](https://github.com/AvaloniaUI/Avalonia/assets/53590491/f88844aa-e1f5-4dc5-932e-35b24d09e6c6)

**Reply by wosiyourfather** [reply]
But still when the data source of my binding changes, the menu will disappear

**Reply by wosiyourfather** [reply]
Dude, are you still there?

**Reply by timunie** [reply]
I'm not "Dude". This is not live support. If you need that, you have to get paid support. 

--------

From what I see I have no more idea. Attach a very minimal sample (no full App!) showing your issue. Then I may take a look if I find the time to do.

**Reply by wosiyourfather** [reply]
> I'm not "Dude". This is not live support. If you need that, you have to get paid support.
> 
> From what I see I have no more idea. Attach a very minimal sample (no full App!) showing your issue. Then I may take a look if I find the time to do.
okey！This is the demo. Could you please help me take a look
[demo.zip](https://github.com/AvaloniaUI/Avalonia/files/12224340/demo.zip)

**Comment by wosiyourfather** [comment]
[demo.zip](https://github.com/AvaloniaUI/Avalonia/files/12224307/demo.zip)
okey！This is the demo. Could you please help me take a look

**Reply by timunie** [reply]
I had a look at your demo. Unfortunately I think it's by design that the context menu disappears if the data itself disappears. You may need to stop the timer once you open the ContextMenu and activate it back after it closes. You can try this: 

```diff
<Border.ContextMenu>
	<ContextMenu  Name="cd"
+	              Opened="Cd_OnOpened"
+	              Closed="Cd_OnClosed">
		<MenuItem Name="cdd" Header="kkxiaoguo"></MenuItem>
		<MenuItem Name="cddw" Header="kkxiaoguo"></MenuItem>
		<MenuItem Name="cddwe" Header="kkxiaoguo"></MenuItem>
	</ContextMenu>
</Border.ContextMenu>
```

in code behind:
```cs
private void Cd_OnOpened(object? sender, RoutedEventArgs e)
{
    // stop timer if menu appears
    if (DataContext is MainViewModel viewModel)
    {
        viewModel._timer.Stop();
    }
}

private void Cd_OnClosed(object? sender, RoutedEventArgs e)
{
    // start timer again
    if (DataContext is MainViewModel viewModel)
    {
        viewModel._timer.Start();
    }
}
```
